src/server/system_services/objects/quota.js

```diff
@@ -0,0 +1,198 @@
+/* Copyright (C) 2021 NooBaa */
+/* eslint max-lines: ['error', 2500] */
+
+/**
+ * Quota wrapper class represents bucket quota configuration and related methods
+ */
+'use strict';
+
+const size_utils = require('../../../util/size_utils');
+const BigInteger = require('big-integer');
+const _ = require('lodash');
+const Dispatcher = require('../../notifications/dispatcher');
+const config = require('../../../../config');
+
+const RAW_VALUE_FIELD_NAME = 'raw_value';
+
+class Quota {
+    /**
+     * 
+     * @param {any} quota_config - bucket quota configuration payload
+     */
+    constructor(quota_config) {
+        this.size = quota_config && quota_config.size ? quota_config.size : {'value': 0};
+         this.size[RAW_VALUE_FIELD_NAME] = size_utils.size_unit_to_bigint(this.size.value, this.size.unit).toJSON();
+         this.quantity = quota_config && quota_config.quantity ? quota_config.quantity : {'value': 0};
+         this.quantity[RAW_VALUE_FIELD_NAME] = new BigInteger(this.quantity.value).toJSON();
+    }
+
+    /**
+     * @returns size value.
+     */
+    get_size_value() {
+        return this.size.value;
+    }
+
+    /**
+     * @returns size unit.
+     */
+    get_size_unit() {
+        return this.size.unit;
+    }
+
+    /**
+     * @returns quantity value.
+     */
+    get_quantity_value() {
+        return this.quantity.value;
+    }
+
+    /**
+     * @returns quantity unit.
+     */
+     get_quantity_unit() {
+        return this.quantity.unit;
+    }
+
+    /**
+     * 
+     * @returns calculated bigint size raw value of quota. default is 0.
+     */
+    get_quota_by_size() {
+        return this.size.raw_value;
+    }
+
+    /**
+     * 
+     * @returns calculated bigint quantity raw value of quota
+     */
+    get_quota_by_quantity() {
+        return this.quantity.raw_value;
+    }
+
+    /**
+     * 
+     * @returns - new quota config object without raw values
+     */
+    get_config() {
+        const quota_config = {};
+        if (this.size.value > 0) {
+            quota_config.size = _.omit(this.size, RAW_VALUE_FIELD_NAME);
+        }
+        if (this.quantity.value > 0) {
+            quota_config.quantity = _.omit(this.quantity, RAW_VALUE_FIELD_NAME);
+        }
+        return quota_config;
+    }
+
+
+  /**
+   * 
+   * @returns the percent of quantity quota used by the bucket. 
+   */
+    static get_bucket_quota_usages_percent(bucket, bucket_quota) {
```

Comment:
why does it need to be static? looks like all the static functions get a quota object so they can be class methods

Reply:
Moved to class